pkgs/build-support/cc-wrapper/default.nix

```diff
@@ -343,10 +344,18 @@ stdenv.mkDerivation {
     # compile, because it uses "#include_next <limits.h>" to find the
     # limits.h file in ../includes-fixed. To remedy the problem,
     # another -idirafter is necessary to add that directory again.
+    #
+    # We use --sysroot=/nix/store/does/not/exist to drop embedded default
+    # path to glibc headers gcc was built against. Without it -idirafter
+    # only appends to the list and outdated glibc headers end up being
+    # used. 'cc-cflags-before' is used to allow user's --sysroot= option
+    # to override our default.
     + optionalString (libc != null) (''
       touch "$out/nix-support/libc-cflags"
       touch "$out/nix-support/libc-ldflags"
       echo "-B${libc_lib}${libc.libdir or "/lib/"}" >> $out/nix-support/libc-crt1-cflags
+    '' + optionalString enableNonExistentSysroot ''
+      echo "--sysroot=/nix/store/does/not/exist" >> $out/nix-support/cc-cflags-before
```

Comment:
As stated in https://github.com/NixOS/nixpkgs/pull/213185 blanket `--sysroot=` changes library search path in transitive libraries. As a result we get linkage failures in `dmd`, `d-seams`, `llvmPackages_rocm.compiler-rt`. Conditional or not it's breaking real packages. The intent in the comment was only attempting to change header search path and not library search path.

For just headers there is an `-isysroot` flag. But it has it's own caveats as blanket `--sysroot=` also exposed failures in `ipxe` and `wimboot`. Those show that existing `-idirafter` sequence is incorrectly overridden in `nixpkgs`. It has to be fixed by chaning a priority.

Reply:
The conditional is only enabled for `xgcc`, which is never used to compile any of those packages.

Reply:
My suggestion would be to explore ways to add the flag locally to the place where it's intended to be used instead of changing the driver. I still think it's too heavy a hammer even for `gcc`.